### PR TITLE
[W07H03] added FieldVisualizeHelper

### PR DESCRIPTION
**Pull request by cod-eric**
added FieldVisualizeHelper, a class that helps to quickly create Cell-Arrays from a textual description, as well as visualize Cell-Arrays textually

Methods are pretty much self-explanatory with their JavaDoc, but here are their descriptions, anyways:

**fieldDescriptionToCellArray**
> Turns a field description into a Cell-Array.
> Note that this method does not support emojis as input (they would be too complicated to type manually, anyways).
> 
> **Field Description Language (FDL)**
> FDL uses the following syntax:
> - `g` = Plant ("Greenery")
> - `h` = Hamster
> - `p` = Pingu
> - `w` = Wolf
> - `.` = Empty Cell
> 
> Cells are separated by spaces, lines should, for better readability, be separated by line breaks, but spaces are fine, too.
> **Example field in FDL**
> ```
> . g g h . h w
> w p . . . . .
> . . . . . . .
> ```
> Alternatively:
> ```
> . g g h . h w w p . . . . . . . . . . . .
> ```
> 
> @param cellsTextual  A String/text block containing the field description in FDL
> @param width           The width of the field
> @param height          The height of the field
> @return                      A Cell-Array with the entities created

**cellArrayToFieldDescription**
> Turns a Cell-Array into a field description.
> 
> **Field Description Language (FDL)**
> FDL uses the following syntax:
> 
> - `g/🌱` = Plant ("Greenery")
> - `h/🐹` = Hamster
> - `p/🐧` = Pingu
> - `w/🐺` = Wolf
> - `./🔳` = Empty Cell
> 
> Cells are separated by spaces, lines will be, for better readability, be separated by line breaks.
> **Example field in FDL**
> ```
>     . g g h . h w
>     w p . . . . .
>     . . . . . . .
> ```
> Alternatively:
> ```
>     🔳 🌱 🌱 🐹 🔳 🐹 🐺
>     🐺 🐧 🔳 🔳 🔳 🔳 🔳
>     🔳 🔳 🔳 🔳 🔳 🔳 🔳
> ```
> 
> @param cellField     A Cell-Array to be turned into an FDL description
> @param width         The width of the field
> @param height        The height of the field
> @param useEmojis     Specifies whether the output should be given as emojis representing the respective cells. Might make it easier to get an overview of the field.
> @return              A String with the textual description of cellField in FDL

### CODE REVIEW

**Comment by Friendly-Banana** [comment]
I have added a [PR](https://github.com/cod-eric/pgdp2223-tests/pull/1) with a correction and suggested changes, otherwise LGTM.

File: w07h03/FieldVisualizeHelper.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
+package pgdp.sim;
```
Comment (jonasfeld) [inline w07h03/FieldVisualizeHelper.java:1]: wrong file path then! imo this should be under `w07h03/test/pgdp/sim`

Comment (jonasfeld) [reply w07h03/FieldVisualizeHelper.java:1]: hm or maybe not test but `helpers` or sth

File: w07h03/FieldVisualizeHelper.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+            } else if (textualSymbols[1].equals(cellFieldTextual[pos])){
+                cellField[pos] = new Hamster();
+            } else if (textualSymbols[2].equals(cellFieldTextual[pos])){
+                cellField[pos] = new Wolf();
+            } else if (textualSymbols[3].equals(cellFieldTextual[pos])){
+                cellField[pos] = new Pingu();
+            } else if (textualSymbols[4].equals(cellFieldTextual[pos])){
+                cellField[pos] = null;
+            }
+        }
```
Comment (jonasfeld) [inline w07h03/FieldVisualizeHelper.java:54]: I really like the pr @Friendly-Banana wrote for this

**Review by jonasfeld — COMMENTED** [review]
This feels very helpful! Thank you for sharing :)

File: w07h03/FieldVisualizeHelper.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+     * @return A String with the textual description of cellField in FDL
+     */
+    public static String cellArrayToFieldDescription(Cell[] cellField, int width, int height, boolean useEmojis) {
+        String[] charSetToUse = useEmojis ? emojiSymbols : textualSymbols;
+
+        StringBuilder sb = new StringBuilder();
+        for (int y = 0; y < height; y++) {
+            for (int x = 0; x < width; x++) {
+                Cell c = cellField[y * width + x];
+                if (c instanceof Plant) {
```
Comment (julian-hartl) [inline w07h03/FieldVisualizeHelper.java:93]: I guess you could simplify that code by using a hash map that maps each class to an index.

**Review by MaximilianAnzinger — APPROVED** [review]
[🐧](https://emojipedia.org/penguin/)

**Comment by jonasfeld** [comment]
As I said, the file path still feels wrong? @Friendly-Banana what's your input on that?

**Comment by Friendly-Banana** [comment]
Should be `w07h03/test/pgdp/sim`.

File: w07h03/FieldVisualizeHelper.java (hunk 1 above)
Comment (cod-eric) [reply w07h03/FieldVisualizeHelper.java:1]: yeah i was looking at my DeGraphViz from W06H03, and there it was this location I think. But can change it if you like.

Comment (jonasfeld) [reply w07h03/FieldVisualizeHelper.java:1]: yes please! @cod-eric

File: w07h03/FieldVisualizeHelper.java (hunk 3 above)
Comment (Friendly-Banana) [reply w07h03/FieldVisualizeHelper.java:93]: Both ways would be cleaner, but break the emoji functionality. Switch requires constant values, so we would have to hardcode chars. We could use two maps, but then need another if statement to choose the right one. Easiest to just leave it as it is.

**Comment by Friendly-Banana** [comment]
I would close this PR, as #133 includes this PR and also fixes the filepath.

**Comment by jonasfeld** [comment]
> I would close this PR, as #133 includes this PR and also fixes the filepath.

Agreed. @MaximilianAnzinger

**Comment by jonasfeld** [comment]
@nilsreichardt ptal

**Pull request closed without merging**